src/psyclone/tests/psyir/nodes/scoping_node_test.py

```diff
@@ -0,0 +1,172 @@
+    inner one. This means that the References to the symbols on the outer
+    scope should not be duplicated. Also it contains argument symbols and
+    a reference inside another reference to make sure all this are copied
+    appropriately.
+    '''
+    parent_node = Container("module")
+    symbol_b = parent_node.symbol_table.new_symbol(
+        "b", symbol_type=DataSymbol, datatype=ArrayType(INTEGER_TYPE, [5]))
+    schedule = Routine("routine")
+    parent_node.addchild(schedule)
+    symbol_a = schedule.symbol_table.new_symbol(
+        "a", symbol_type=DataSymbol, datatype=INTEGER_TYPE,
+        interface=ArgumentInterface(ArgumentInterface.Access.READWRITE))
+    schedule.symbol_table.specify_argument_list([symbol_a])
+    symbol_i = schedule.symbol_table.new_symbol(
+        "i", symbol_type=DataSymbol, datatype=INTEGER_TYPE)
+
+    schedule.addchild(
+        Assignment.create(Reference(symbol_a),
+                          ArrayReference.create(symbol_b,
+                                                [Reference(symbol_i)])))
+
+    new_schedule = schedule.copy()
+
+    # Check that the symbol_table has been deep copied
+    assert new_schedule.symbol_table is not schedule.symbol_table
+    assert new_schedule.symbol_table.lookup("i") is not \
+        schedule.symbol_table.lookup("i")
+    assert new_schedule.symbol_table.lookup("a") is not \
+        schedule.symbol_table.lookup("a")
+
+    # Check that 'a' and 'i' have been copied to the new symbol table.
+    assert new_schedule[0].lhs.symbol not in schedule.symbol_table.symbols
+    assert new_schedule[0].lhs.symbol in new_schedule.symbol_table.symbols
+    assert new_schedule[0].rhs.children[0].symbol not in \
+        schedule.symbol_table.symbols
+    assert new_schedule[0].rhs.children[0].symbol in \
+        new_schedule.symbol_table.symbols
+
+    # Add the "_new" suffix to all symbol in the copied schedule
+    for symbol in new_schedule.symbol_table.symbols:
+        new_schedule.symbol_table.rename_symbol(symbol, symbol.name+"_new")
+
+    # An update to a symbol to the outer scope must affect both copies of the
+    # inner schedule.
+    parent_node.symbol_table.rename_symbol(symbol_b, symbol_b.name+"_global")
+
+    # Insert the schedule back to the original container
+    parent_node.addchild(new_schedule)
+
+    # Check that the expected code is generated
+    # TODO #1200: the new 'routine' RoutineSymbol also needs to change.
+    expected = '''\
+module module
+  implicit none
+  integer, dimension(5) :: b_global
+
+  contains
+  subroutine routine(a)
+    integer, intent(inout) :: a
+    integer :: i
+
+    a = b_global(i)
+
+  end subroutine routine
+  subroutine routine(a_new)
+    integer, intent(inout) :: a_new
+    integer :: i_new
+
+    a_new = b_global(i_new)
+
+  end subroutine routine
+
+end module module'''
+    writer = FortranWriter()
+    assert expected in writer(parent_node)
```

Comment:
If, once #1200 is done, this could be compiled then please add a TODO for that here.

Reply:
Done